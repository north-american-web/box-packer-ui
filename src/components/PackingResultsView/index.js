import React from 'react'
import PropTypes from "prop-types";
import {Panel, Toast} from "../Spectre";
import PackingResultsListView from '../PackingResultsListView/index';

function PackingResultsView({success, packed = [], empty = [], leftOverItems = [], apiRequest, apiResponse}) {
    return (
        <Panel title='API Response'>
            <Toast status={success ? 'success' : 'error'}>
                The item(s) {!success && 'won\'t '}fit into the box(es)!
            </Toast>

            <div className="columns mt-2">
                <div className="column col-md-12 col-4">
                    <PackingResultsListView
                        title='Packed boxes'
                        solids={packed}
                    />
                </div>

                <div className="column col-md-12 col-4">
                    <PackingResultsListView
                        title='Empty boxes'
                        solids={empty}
                    />
                </div>

                <div className="column col-md-12 col-4">
                    <PackingResultsListView
                        title='Left-over items'
                        solids={leftOverItems}
                        fallbackDescription='Item'
                    />
                </div>
            </div>

            <h6>Request JSON</h6>
            <pre className="code" data-lang='JSON'><code data-testid="request-json">{JSON.stringify(apiRequest, null, 2)}</code></pre>

            <h6>Response JSON</h6>
            <pre className="code" data-lang='JSON'><code data-testid="response-json">{JSON.stringify(apiResponse, null, 2)}</code></pre>
        </Panel>
    )
}

PackingResultsView.propTypes = {
    success: PropTypes.bool.isRequired,
    packed: PropTypes.array,
    empty: PropTypes.array,
    leftOverItems: PropTypes.array,
    apiRequest: PropTypes.object.isRequired,
    apiResponse: PropTypes.object.isRequired
};

export default PackingResultsView;