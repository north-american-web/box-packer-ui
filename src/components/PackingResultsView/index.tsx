import React from 'react'
import PropTypes from "prop-types";
import {Panel} from "../Spectre";
import {SolidInterface, SolidView} from "../Solid";
import {PackApiRequest, PackApiResponse} from "../../utils/boxPackerAPI";

interface PackingResultsListViewProps {
    title: string;
    solids: SolidInterface[];
    fallbackDescription?: string;
}

export const PackingResultsListView = ({title, solids, fallbackDescription = 'Box'}: PackingResultsListViewProps) => {
    return (
        <>
            <h6 data-testid="packing-results-list-view__title">{title}</h6>
            {solids.length === 0
                ? (
                    <p data-testid="packing-results-list-view__contents" className="text-gray">Nothing here.</p>
                )
                : (
                    <ul data-testid="packing-results-list-view__contents">
                        {solids.map((solid, index) => (
                            <li key={index}>
                                <SolidView fallbackDescription={fallbackDescription} solid={solid}/>
                            </li>
                        ))}
                    </ul>
                )}
        </>
    )
};
PackingResultsListView.propTypes = {
    title: PropTypes.string.isRequired,
    solids: PropTypes.array.isRequired
};

export interface PackingResultsViewProps {
    apiLoadingTime: number;
    apiRequest: PackApiRequest;
    apiResponse: PackApiResponse;
}

function PackingResultsView({apiLoadingTime, apiRequest, apiResponse}: PackingResultsViewProps) {
    return (
        <Panel title={(
            <>
                Last API Response <small>(Fetched in <span data-testid="loading-time">{apiLoadingTime}</span>ms)</small>
            </>
        )}>
            <div className="columns mt-2">
                <div className="column col-md-12 col-4">
                    <PackingResultsListView
                        title='Packed boxes'
                        solids={apiResponse.packed}
                    />
                </div>

                <div className="column col-md-12 col-4">
                    <PackingResultsListView
                        title='Empty boxes'
                        solids={apiResponse.empty}
                    />
                </div>

                <div className="column col-md-12 col-4">
                    <PackingResultsListView
                        title='Left-over items'
                        solids={apiResponse.leftOverItems}
                        fallbackDescription='Item'
                    />
                </div>
            </div>

            <div className="columns mt-2">
                <div className="column col-md-12 col-6">
                    <h6>Request JSON</h6>
                    <pre className="code" data-lang='JSON'><code
                        data-testid="request-json">{JSON.stringify(apiRequest, null, 2)}</code></pre>
                </div>
                <div className="column col-md-12 col-6">
                    <h6>Response JSON</h6>
                    <pre className="code" data-lang='JSON'><code
                        data-testid="response-json">{JSON.stringify(apiResponse, null, 2)}</code></pre>
                </div>
            </div>
        </Panel>
    )
}

PackingResultsView.propTypes = {
    apiLoadingTime: PropTypes.number.isRequired,
    apiRequest: PropTypes.object.isRequired,
    apiResponse: PropTypes.object.isRequired
};

export default PackingResultsView;