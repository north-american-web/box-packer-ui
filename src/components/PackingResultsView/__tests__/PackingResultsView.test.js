import React from 'react';
import 'jest-dom/extend-expect';
import { render, cleanup } from '@testing-library/react';
import PackingResultsView from '../index'

afterEach(cleanup);

describe('<PackingResultsView/>', () => {
    it('renders and displays correctly', () => {
        const packed = [ {width: 1, length: 2, height: 3}];
        const empty = [ {width: 4, length: 5, height: 6}];
        const leftOverItems = [{width: 7, length: 8, height: 9}];
        const apiRequest = {label: 'api-request'};
        const apiResponse = {label: 'api-response'};

        const renderResult = render(<PackingResultsView
                                        success={true}
                                        empty={empty}
                                        packed={packed}
                                        leftOverItems={leftOverItems}
                                        apiResponse={apiResponse}
                                        apiRequest={apiRequest} />);
        expect(renderResult.queryByText('The item(s) fit into the box(es)!')).not.toBeNull();
        expect(renderResult.getByTestId('toast-element')).toHaveClass('toast-success');

        const titles = renderResult.queryAllByTestId('packing-results-list-view__title');
        expect(titles[0]).toHaveTextContent('Packed boxes');
        expect(titles[1]).toHaveTextContent('Empty boxes');
        expect(titles[2]).toHaveTextContent('Left-over items');

        const contents = renderResult.queryAllByTestId('packing-results-list-view__contents');
        expect(contents[0].children.length).toBe(1);
        expect(contents[1].children.length).toBe(1);
        expect(contents[2].children.length).toBe(1);

        expect(renderResult.getByTestId('request-json')).toMatchSnapshot();
        expect(renderResult.getByTestId('response-json')).toMatchSnapshot();
    });

    it('handles packing failure and empty values correctly', () => {
        const renderResult = render(<PackingResultsView
            success={false}
            apiResponse={{label: 'api-request'}}
            apiRequest={{label: 'api-response'}} />);

        expect(renderResult.queryByText('The item(s) won\'t fit into the box(es)!')).not.toBeNull();

        const contents = renderResult.queryAllByTestId('packing-results-list-view__contents');
        expect(contents[0].children.length).toBe(0);
        expect(contents[1].children.length).toBe(0);
        expect(contents[2].children.length).toBe(0);
    })

});