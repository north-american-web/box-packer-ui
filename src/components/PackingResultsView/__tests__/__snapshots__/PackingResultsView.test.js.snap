// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PackingResultsView/> renders and displays correctly 1`] = `
<code
  data-testid="request-json"
>
  {
  "label": "api-request"
}
</code>
`;

exports[`<PackingResultsView/> renders and displays correctly 2`] = `
<code
  data-testid="response-json"
>
  {
  "label": "api-response"
}
</code>
`;
